### PR TITLE
Flo data input

### PR DESCRIPTION
**Pull request by floschne**
**Implemented basic functionality to load Wikipedia Articles.**
I think the tests in `WikiHttpApiLoaderTest` are good entry points to review / understand the code. Basically I'm firing a query to the Wikipedia API via HTTP GET and filter the response. The result  is an instance of a `WikiArticle` containing the content of the stripped response and representing an Wikipedia Article.

### CODE REVIEW

File: dev/src/main/java/dataInput/JsoupWikiLoader.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
+package dataInput;
+
+import java.util.MissingResourceException;
+import java.util.Set;
+
+/**
+ * This class downloads @{@link WikiArticle}s via JSOUP library
+ */
+public class JsoupWikiLoader implements WikiArticleLoader {
```
Comment (julianbetz) [inline dev/src/main/java/dataInput/JsoupWikiLoader.java]: Is this a stub class or a piece of code not intended to be part of the final version?

File: dev/src/main/java/dataInput/WikiHttpApiLoader.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+ * <p>
+ * Example API call: https://de.wikipedia.org/w/api.php?action=query&prop=extracts&explaintext&format=xml&titles=Pizza
+ */
+public class WikiHttpApiLoader implements WikiArticleLoader {
+
+    private static final String WIKI_API_BASE_URL_LANGUAGE_TOKEN = "<lang>";
+    private static final String WIKI_API_BASE_URL_TITLE_TOKEN = "<title>";
+    private static final String WIKI_API_BASE_URL_FORMAT_TOKEN = "<format>";
+    private static final String DEFAULT_FORMAT = "xml";
+    private static final String API_RESPONSE_HEADING_REGEX_PATTERN = "={2,4} [.\\w\\s]+ ={2,4}";
```
Comment (julianbetz) [inline dev/src/main/java/dataInput/WikiHttpApiLoader.java]: The headings possibly include punctuation marks as well, espc. parentheses.

File: dev/src/main/java/dataInput/WikiArticleReader.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    //TODO make this mandatory and with UIMA parameter to create enginge..
+    private List<Pair<WikiArticle.Language, String>> wikiContentToProcess; //TODO better name?
+
+    public WikiArticleReader() {
+        this.wikiArticleLoader = WikiHttpApiLoader.getInstance();
+    }
+
+    public WikiArticleReader(WikiArticleLoader loader) {
+        if(loader == null)
+            throw new IllegalArgumentException("WikiArticleLoader must not be null or empty!");
```
Comment (julianbetz) [inline dev/src/main/java/dataInput/WikiArticleReader.java]: I guess you mean simply "... must not be null."

File: dev/src/main/java/dataInput/WikiArticleReader.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    /**
+     * Subclasses should implement this method rather than {@link #getNext(CAS)}
+     *
+     * @param jCas the {@link JCas} to store the read data to
+     * @throws IOException         if there was a low-level I/O problem
+     * @throws CollectionException if there was another problem
+     */
+    @Override
+    public void getNext(JCas jCas) throws IOException, CollectionException {
+
```
Comment (julianbetz) [inline dev/src/main/java/dataInput/WikiArticleReader.java]: Still a stub, I assume. The same seems to be the case for hasNext() and getProgress().

File: dev/src/main/java/dataInput/WikiArticleReader.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+/**
+ * Reads @{@link WikiArticle}s and adds it to the JCas
+ */
+public class WikiArticleReader extends JCasCollectionReader_ImplBase {
+
+    //the WikiArticleLoader that'll be used to provide the articles
+    private WikiArticleLoader wikiArticleLoader;
+    //will contain the articles that are already loaded
+    private List<WikiArticle> wikiArticles;
```
Comment (julianbetz) [inline dev/src/main/java/dataInput/WikiArticleReader.java]: Do you intend to store all fetched articles here? It seems to be sensible to get an article from Wikipedia not before the preceding ones are processed and the next one is requested via getNext(). This way, every article can be worked on individually in the pipeline, saving storage.

File: dev/src/main/java/dataInput/WikiArticleLoader.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    WikiArticle loadArticle(String title, WikiArticle.Language language) throws MissingResourceException, IOException;
+
+
+    /**
+     * Downloads the articles from Wikipedia with title @title in the languages @languages and returns it as a set of @{@link WikiArticle}
+     * @param title the title of the article
+     * @param languages the languages of the article
+     * @return a set containing the @{@link WikiArticle}s. If an article cannot be found in the specific language it'll be ignored
+     */
+    Set<WikiArticle> loadArticleInMultipleLanguages(String title, Set<WikiArticle.Language> languages);
```
Comment (julianbetz) [inline dev/src/main/java/dataInput/WikiArticleLoader.java]: I expect this to work better or worse depending on how you will implement it: If you search for the same title in different languages, you will undoubtedly get a lot of misses save the cases in which titles are named entities; if you do the equivalent of following the language links on a Wikipedia HTML page, you might have a chance to get a lot of articles in different languages. I find it difficult to estimate however how feasible that second option really is.

**Review by julianbetz — APPROVED** [review]
All in all okay, but WikiArticleReader will not work in a pipeline in this stage of development. Is this what you intended?
Please also mark methods/classes that are not yet implemented with "// TODO" or something similar so no one starts writing code that relies on these parts.
Concerning the Codacy check: Since we are using a style checker, you might want to address some code style issues.

File: dev/src/main/java/dataInput/JsoupWikiLoader.java (hunk 1 above)
Comment (floschne) [reply dev/src/main/java/dataInput/JsoupWikiLoader.java]: was just an alternative idea - now deleted

File: dev/src/main/java/dataInput/WikiArticleLoader.java (hunk 6 above)
Comment (floschne) [reply dev/src/main/java/dataInput/WikiArticleLoader.java]: true point! now deleted

File: dev/src/main/java/dataInput/WikiArticleReader.java (hunk 3 above)
Comment (floschne) [reply dev/src/main/java/dataInput/WikiArticleReader.java]: edited

File: dev/src/main/java/dataInput/WikiArticleReader.java (hunk 4 above)
Comment (floschne) [reply dev/src/main/java/dataInput/WikiArticleReader.java]: have to talk about this in next meeting :)

File: dev/src/main/java/dataInput/WikiArticleReader.java (hunk 5 above)
Comment (floschne) [reply dev/src/main/java/dataInput/WikiArticleReader.java]: have to talk about this in next meeting :)

**Comment by floschne** [comment]
Waiting for merge until next meeting :)

**Pull request merged**